File: 02-10/6/ac/solution.py
class Node():
	def __init__(self, value):
		self.next = None
		self.previous = None
		self.value = value

	def __str__(self):
		ans = 'NEXT IS ' + str(self.next.value)
		ans = ans + '\nPREVIOUS IS ' + str(self.previous.value) + '\n'
		return ans

class Buffer():

	def __init__(self):
		self.head = Node('HEAD')
		self.tail = Node('TAIL')
		self.head.next = self.tail
		self.tail.previous = self.head 
		self.storage = {}

	def contains(self, value):
		return value in self.storage

	def addFront(self, value):
		newNode = Node(value)
		self.storage[value] = newNode
		newNode.previous = self.head
		newNode.next = self.head.next
		self.head.next.previous = newNode
		self.head.next = newNode

	def remove(self, value):
		toRemove = self.storage[value]
		toRemove.previous.next = toRemove.next
		toRemove.next.previous = toRemove.previous
		del self.storage[value]

	def removeFirst(self):
		if (len(self.storage) == 0):
			return None
		returnVal = self.head.next.value
		self.remove(returnVal)
		return returnVal

	def printList(self):
		curr = self.head.next
		while (curr != self.tail):
			print curr.value
			curr = curr.next
		print ''

def main():
	cases = int(raw_input())
	for case in range(cases):
		commands = int(raw_input())
		container = Buffer()
		for j in range(commands):
			line = raw_input()
			if (line == 'DELETE'):
				val = container.removeFirst()
				if (val == None):
					print 'NONE'
				else:
					print val
			else:
				action, person = line.split()
				if (action != 'SEND' and action != 'RECEIVE'):
					raise Exception('Unexpected value ' + action)
				if (container.contains(person)):
					container.remove(person)
				container.addFront(person)
		print('')

main()

File: 02-10/6/ac/Solution.java
import java.util.*;
import java.io.*;

public class Solution {

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int cases = Integer.parseInt(br.readLine());
		StringBuilder sb = new StringBuilder();
		for (int a = 0; a < cases; a++) {
			int interactions = Integer.parseInt(br.readLine());
			Buffer other = new Buffer();
			for (int b = 0; b < interactions; b++) {
				StringTokenizer st = new StringTokenizer(br.readLine());
				String command = st.nextToken();
				if (command.equals("SEND") || command.equals("RECEIVE")) {
					String name = st.nextToken();
					if (other.contains(name))
						other.remove(name);
					other.addFront(name);
				} else if (command.equals("DELETE")) {
					String result = other.removeFront();
					if (result == null)
						sb.append("NONE\n");
					else {
						sb.append(result);
						sb.append('\n');
					}
				} else {
					throw new IllegalArgumentException("Didn't recognize command. " + command);
				}
			}
			sb.append('\n');
		}
		System.out.print(sb.toString());
	}

	private static class Buffer {

		private static class Node {
			private Node previous;
			private Node next;
			private String value;

			public Node(Node previous, Node next, String value) {
				this.previous =  previous;
				this.next = next;
				this.value = value;
			}
		}

		private Node head;
		private Node tail;
		HashMap<String, Node> map;

		public Buffer() {
			head = new Node(null, null, null);
			tail = new Node(null, null, null);
			head.next = tail;
			tail.previous = head;
			map = new HashMap<String, Node>();
		}

		public boolean contains(String name) {
			return map.containsKey(name);
		}

		public void remove(String name) {
			Node tgt = map.get(name);
			tgt.next.previous = tgt.previous;
			tgt.previous.next = tgt.next;
			map.remove(name);
		}

		public String removeFront() {
			if (head.next == tail)
				return null;
			String toRemove = head.next.value;
			remove(toRemove);
			return toRemove;
		}

		public void addFront(String name) {
			Node n = new Node(head, head.next, name);
			n.next.previous = n;
			head.next = n;
			map.put(name, n);
		}

		// DEBUGGING PURPOSES ONLY
		public String getPrint() {
			StringBuilder sb = new StringBuilder();
			Node curr = head.next;
			while (curr != tail) {
				sb.append(curr.value);
				curr = curr.next;
				sb.append('\n');
			}
			return sb.toString();
		}
	}
}